pkg/services/sqlstore/migrator/mysql_dialect.go

```diff
@@ -225,3 +229,76 @@ func (db *MySQLDialect) UpsertSQL(tableName string, keyCols, updateCols []string
 	)
 	return s
 }
+
+func (db *MySQLDialect) Lock() error {
+	sess := db.engine.NewSession()
+	defer sess.Close()
+
+	query := "SELECT GET_LOCK(?, ?)"
+	var success bool
+
+	lockName, err := db.getLockName()
+	if err != nil {
+		return fmt.Errorf("failed to generate lock name: %w", err)
+	}
+
+	// trying to obtain the lock with the specific name
+	// the lock is exclusive per session and is released explicitly by executing RELEASE_LOCK() or implicitly when the session terminates
+	// it returns 1 if the lock was obtained successfully,
+	// 0 if the attempt timed out (for example, because another client has previously locked the name),
+	// or NULL if an error occurred
+	// starting from MySQL 5.7 it is even possible for a given session to acquire multiple locks for the same name
+	// however other sessions cannot acquire a lock with that name until the acquiring session releases all its locks for the name.
+	_, err = sess.SQL(query, lockName, 0).Get(&success)
+	if err != nil {
+		return err
+	}
+	if !success {
+		return ErrLockDB
+	}
+	return nil
+}
+
+func (db *MySQLDialect) Unlock() error {
+	sess := db.engine.NewSession()
+	defer sess.Close()
+
+	query := "SELECT RELEASE_LOCK(?)"
+	var success bool
+
+	lockName, err := db.getLockName()
+	if err != nil {
+		return fmt.Errorf("failed to generate lock name: %w", err)
+	}
+
+	// trying to release the lock with the specific name
+	// it returns 1 if the lock was released,
+	// 0 if the lock was not established by this thread (in which case the lock is not released),
+	// and NULL if the named lock did not exist (it was never obtained by a call to GET_LOCK() or if it has previously been released)
+	_, err = sess.SQL(query, lockName).Get(&success)
+	if err != nil {
+		return err
+	}
+	if !success {
+		return ErrReleaseLockDB
+	}
+	return nil
+}
+
+func (db *MySQLDialect) getLockName() (string, error) {
+	cfg, err := mysql.ParseDSN(db.engine.DataSourceName())
+	if err != nil {
+		return "", err
+	}
+
+	s, err := database.GenerateAdvisoryLockId(cfg.DBName)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate advisory lock key: %w", err)
+	}
+
+	if len(s) > max_allowed_lockname_length {
+		// what if there several databases with same prefix of 64 length or more?
```

Comment:
We need to document this as a potential breaking change at least. Random thought, might be able to use https://grafana.com/docs/grafana/latest/administration/configuration/#instance_name as fallback, but not sure since it can be empty.

Reply:
> We need to document this as a potential breaking change at least.
do you mean not allowing having databases with longer names?

> Random thought, might be able to use https://grafana.com/docs/grafana/latest/administration/configuration/#instance_name as fallback, but not sure since it can be empty.
but still there is not any limit on the  length of `instance_name`, right?

Reply:
> do you mean not allowing having databases with longer names?

Yes. Just in case someone hits this we should have it covered under breaking changes in release notes at least. Question is what to suggest if this happens, rename the database? Might not be possible/hard. Do we need a config to fallback on perhaps? I really like to keep the number of configuration parameters as low as possible so see this as a last resort so lets wait with this and see if someone hits this problem.

> but still there is not any limit on the length of instance_name, right?

True. Forgot my suggestion since instance_name is supposed to be/is probably unique :smile:

Reply:
checking again: this is [crc32](https://github.com/golang-migrate/migrate/blob/3e43c429ee00336a1e8e3bc1f9a855d55b3b4414/database/util.go#L17) so it's length it's should be fixed.
I tend to believe that this redundant.

Reply:
You mean `if len(s) > max_allowed_lockname_length {` is redundant?

Reply:
I also think this is redundant. IIUC: CRC checksum is 32 bits, with a theoretical max value of 2147483647. 2147483647 * (the salt) 1486364155 = 3191942716349473285, which as a string has length 19.